File: Samples/DeviceEnumerationAndPairing/js/js/scenario3_backgrounddevicewatcher.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var DevEnum = Windows.Devices.Enumeration;
    var backgroundTaskName = "DeviceEnumerationJs_BackgroundTaskName";
    var backgroundTaskRegistration = null;

    var page = WinJS.UI.Pages.define("/html/scenario3_backgrounddevicewatcher.html", {
        ready: function (element, options) {

            // Hook up button event handlers
            document.getElementById("startWatcherButton").addEventListener("click", startWatcher, false);
            document.getElementById("stopWatcherButton").addEventListener("click", stopWatcher, false);

            // Manually bind selector options
            DisplayHelpers.backgroundDeviceWatcherSelectors.forEach(function each(item) {
                var option = document.createElement("option");
                option.textContent = item.displayName;
                selectorComboBox.appendChild(option);
            });

            // Process any data bindings
            WinJS.UI.processAll();

            // Determine if the background task is already active
            var allTasks = Windows.ApplicationModel.Background.BackgroundTaskRegistration.allTasks;

            var iter = allTasks.first();
            while (iter.hasCurrent) {
                var task = iter.current.value;

                if (task.name === backgroundTaskName) {
                    backgroundTaskRegistration = task;
                    startWatcherButton.disabled = true;
                    stopWatcherButton.disabled = false;
                    break;
                }

                iter.moveNext();
            }
        }
    });

    function startWatcher() {
        var deviceWatcher = null;
        startWatcherButton.disabled = true;

        // First get the device selector chosen by the UI.
        var selectedItem = DisplayHelpers.backgroundDeviceWatcherSelectors.getAt(selectorComboBox.selectedIndex);

        if (null != selectedItem.deviceClassSelector) {
            // If the a pre-canned device class selector was chosen, call the DeviceClass overloa
            deviceWatcher = DevEnum.DeviceInformation.createWatcher(selectedItem.deviceClassSelector);
        }
        else if (selectedItem.kind == DevEnum.DeviceInformationKind.unknown) {
            // Use AQS string selector from dynamic call to a device api's GetDeviceSelector call
            // Kind will be determined by the selector
            deviceWatcher = DevEnum.DeviceInformation.createWatcher(
                selectedItem.selector,
                null // don't request additional properties for this sample
                );
        }
        else {
            // Kind is specified in the selector info
            deviceWatcher = DevEnum.DeviceInformation.createWatcher(
                selectedItem.selector,
                null, // don't request additional properties for this sample
                selectedItem.kind);
        }

        // Get the background trigger for this watcher
        var triggerEventKinds = [DevEnum.DeviceWatcherEventKind.add, DevEnum.DeviceWatcherEventKind.update, DevEnum.DeviceWatcherEventKind.remove];
        var deviceWatcherTrigger = deviceWatcher.getBackgroundTrigger(triggerEventKinds);
        WinJS.log && WinJS.log("Starting watcher...", "sample", "status");
        
        // Register this trigger for our background task
        registerBackgroundTask(deviceWatcherTrigger);

        stopWatcherButton.disabled = false;
        WinJS.log && WinJS.log("Watcher started...", "sample", "status");
    }

    function stopWatcher() {

        stopWatcherButton.disabled = true;
        if (null != backgroundTaskRegistration) {
            unregisterBackgroundTask(backgroundTaskRegistration);
        }
        startWatcherButton.disabled = false;
    }

    function registerBackgroundTask(deviceWatcherTrigger) {

        // First see if we already have this background task registered. If so, unregister
        // it before we register it with the new trigger.
        var allTasks = Windows.ApplicationModel.Background.BackgroundTaskRegistration.allTasks;

        var iter = allTasks.first();
        while (iter.hasCurrent) {
            var task = iter.current.value;

            if (task.name === backgroundTaskName) {
                unregisterBackgroundTask(task);
                break;
            }

            iter.moveNext();
        }

        var taskBuilder = new Windows.ApplicationModel.Background.BackgroundTaskBuilder();
        taskBuilder.name = backgroundTaskName;
        taskBuilder.taskEntryPoint = "js\\BackgroundDeviceWatcher.js";
        taskBuilder.setTrigger(deviceWatcherTrigger);

        backgroundTaskRegistration = taskBuilder.register();
        backgroundTaskRegistration.addEventListener("completed", onCompleted)
    }

    function onCompleted(args) {
        WinJS.log && WinJS.log("Background task completed", "sample", "status");
    }

    function unregisterBackgroundTask(taskReg) {
        taskReg.unregister(true);
        var settings = Windows.Storage.ApplicationData.current.localSettings;
        settings.values["eventCount"] = 0;
    }

})();

File: Samples/DeviceEnumerationAndPairing/cppwinrt/Scenario3_BackgroundDeviceWatcher.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario3_BackgroundDeviceWatcher.h"
#include "SampleConfiguration.h"
#include "DeviceHelpers.h"
#include "DeviceWatcherHelper.h"
#include "Scenario3_BackgroundDeviceWatcher.g.cpp"

using namespace winrt;
using namespace winrt::Windows::ApplicationModel::Background;
using namespace winrt::Windows::Devices::Enumeration;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::Storage;
using namespace winrt::Windows::UI::Xaml;
using namespace winrt::Windows::UI::Xaml::Navigation;

winrt::hstring backgroundTaskName = L"DeviceEnumeration_BackgroundTaskName";

namespace winrt::SDKTemplate::implementation
{
    Scenario3_BackgroundDeviceWatcher::Scenario3_BackgroundDeviceWatcher()
    {
        InitializeComponent();
    }

    void Scenario3_BackgroundDeviceWatcher::OnNavigatedTo(NavigationEventArgs const&)
    {
        selectorComboBox().ItemsSource(DeviceSelectorChoices::BackgroundDeviceWatcherSelectors());
        selectorComboBox().SelectedIndex(0);

        // Determine if the background task is already active
        for (auto&& task : BackgroundTaskRegistration::AllTasks())
        {
            if (backgroundTaskName == task.Value().Name())
            {
                backgroundTaskRegistration = task.Value();
                startWatcherButton().IsEnabled(false);
                stopWatcherButton().IsEnabled(true);
                break;
            }
        }
    }

    void Scenario3_BackgroundDeviceWatcher::OnNavigatedFrom(NavigationEventArgs const&)
    {
        if (backgroundTaskRegistration != nullptr)
        {
            backgroundTaskRegistration.Completed(taskCompletedToken);
            backgroundTaskRegistration = nullptr;
        }
    }

    void Scenario3_BackgroundDeviceWatcher::StartWatcherButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        StartWatcher();
    }

    void Scenario3_BackgroundDeviceWatcher::StopWatcherButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        StopWatcher();
    }

    void Scenario3_BackgroundDeviceWatcher::StartWatcher()
    {
        std::vector<DeviceWatcherEventKind> triggerEventKinds = { DeviceWatcherEventKind::Add, DeviceWatcherEventKind::Remove, DeviceWatcherEventKind::Update };

        startWatcherButton().IsEnabled(false);

        // First get the device selector chosen by the UI.
        IInspectable item = selectorComboBox().SelectedItem();
        DeviceSelectorInfo* deviceSelectorInfo = FromInspectable<DeviceSelectorInfo>(item);

        DeviceWatcher deviceWatcher{ nullptr };
        if (deviceSelectorInfo->Selector().empty())
        {
            // If a pre-canned device class selector was chosen, call the DeviceClass overload
            deviceWatcher = DeviceInformation::CreateWatcher(deviceSelectorInfo->DeviceClassSelector());
        }
        else if (deviceSelectorInfo->Kind() == DeviceInformationKind::Unknown)
        {
            // Use AQS string selector from dynamic call to a device api's GetDeviceSelector call
            // Kind will be determined by the selector
            deviceWatcher = DeviceInformation::CreateWatcher(
                deviceSelectorInfo->Selector(),
                nullptr // don't request additional properties for this sample
            );
        }
        else
        {
            // Kind is specified in the selector info
            deviceWatcher = DeviceInformation::CreateWatcher(
                deviceSelectorInfo->Selector(),
                nullptr, // don't request additional properties for this sample
                deviceSelectorInfo->Kind());
        }

        //
        // Get the background trigger for this watcher
        DeviceWatcherTrigger deviceWatcherTrigger = deviceWatcher.GetBackgroundTrigger(triggerEventKinds);

        rootPage.NotifyUser(L"Starting Watcher...", NotifyType::StatusMessage);

        // Register this trigger for our background task
        RegisterBackgroundTask(deviceWatcherTrigger);

        stopWatcherButton().IsEnabled(true);
    }

    void Scenario3_BackgroundDeviceWatcher::RegisterBackgroundTask(DeviceWatcherTrigger const& deviceWatcherTrigger)
    {
        // First see if we already have this background task registered. If so, unregister
        // it before we register it with the new trigger.

        for (auto&& task : BackgroundTaskRegistration::AllTasks())
        {
            if (backgroundTaskName == task.Value().Name())
            {
                UnregisterBackgroundTask(task.Value());
            }
        }

        BackgroundTaskBuilder taskBuilder;
        taskBuilder.Name(backgroundTaskName);
        taskBuilder.TaskEntryPoint(L"BackgroundDeviceWatcherTask.BackgroundDeviceWatcher");
        taskBuilder.SetTrigger(deviceWatcherTrigger);

        backgroundTaskRegistration = taskBuilder.Register();
        taskCompletedToken = backgroundTaskRegistration.Completed({ get_weak(), &Scenario3_BackgroundDeviceWatcher::OnTaskCompleted });
    }

    fire_and_forget Scenario3_BackgroundDeviceWatcher::OnTaskCompleted(BackgroundTaskRegistration const&, BackgroundTaskCompletedEventArgs const&)
    {
        co_await resume_foreground(Dispatcher());

        rootPage.NotifyUser(L"Background task completed", NotifyType::StatusMessage);
    }

    void Scenario3_BackgroundDeviceWatcher::StopWatcher()
    {
        stopWatcherButton().IsEnabled(false);
        if (backgroundTaskRegistration != nullptr)
        {
            UnregisterBackgroundTask(backgroundTaskRegistration);
        }
        startWatcherButton().IsEnabled(true);
    }

    void Scenario3_BackgroundDeviceWatcher::UnregisterBackgroundTask(IBackgroundTaskRegistration const& taskReg)
    {
        taskReg.Unregister(true);
        ApplicationDataContainer settings = ApplicationData::Current().LocalSettings();
        settings.Values().Insert(L"eventCount", box_value(0U));
    }
}
